### PR TITLE
WIP: virtual networking for Docker containers

### PR DESCRIPTION
**Pull request by jdef**
- docker.portMappings.hostPort is optional
- support for USER network mode in Marathon scala API
- add USER as a valid networking mode to docker.network JSON schema
- portMappings require hostPort in docker BRIDGE network mode
- move container portMappings, hostPorts, servicePorts from AppDefinition to Container
- marathon serialization proto tests for docker networking
- disallow ipaddress/discoveryInfo for docker containers using BRIDGE/USER networking
- updated docs for USER mode and portMapping
- generate PORT{x} and PORT_{y} variables more consistently
- removed unused portIndex from PortAssignment

Conflicts:
	src/main/java/mesosphere/marathon/Protos.java
	src/main/scala/mesosphere/mesos/TaskBuilder.scala
	src/test/scala/mesosphere/mesos/TaskBuilderTest.scala

### CODE REVIEW

File: src/main/scala/mesosphere/mesos/TaskBuilder.scala [hunk 1]
```diff
@@ -342,33 +346,83 @@ object TaskBuilder {
...
+          val generated = gp(n)
+          env += (s"PORT_$generated" -> generated.toString)
+      }
+
       portNames.zip(assignedPorts).foreach {
-        case (Some(portName), assignedPort) =>
+        case (Some(portName), Some(assignedPort)) =>
           env += (s"PORT_${portName.toUpperCase}" -> assignedPort.toString)
-        case (None, _) =>
+        // TODO(jdef) port name envvars for generated container ports
```
Comment (jdef) [inline src/main/scala/mesosphere/mesos/TaskBuilder.scala]: TODO

File: src/main/scala/mesosphere/mesos/TaskBuilder.scala [hunk 2]
```diff
@@ -342,33 +346,83 @@ object TaskBuilder {
...
       Map.empty
     }
     else {
       val env = Map.newBuilder[String, String]
+      val generatedPorts = Map.newBuilder[Int, Int] // index -> container port
+
+      object ContainerPortGenerator {
+        // track which port numbers are already referenced by PORT_xxx envvars
+        lazy val consumedPorts = scala.collection.mutable.Set(definedPorts: _*) ++= assignedPorts.flatten
+        lazy val rand = new scala.util.Random(System.currentTimeMillis)
```
Comment (jasongilanfarr) [inline src/main/scala/mesosphere/mesos/TaskBuilder.scala]: Probably want to just use "Random" instead of seeding it (it should be seeded automatically)

File: src/main/scala/mesosphere/mesos/TaskBuilder.scala [hunk 3]
```diff
@@ -342,33 +346,83 @@ object TaskBuilder {
...
+        val maxPort: Int = 65535 - 1024
+
+        // carefully pick a container port that doesn't overlap with other ports used by this
+        // container. and avoid ports in the range (0 - 1024)
+        def next: Int = {
+          val p = rand.nextInt(maxPort) + 1025
```
Comment (jasongilanfarr) [inline src/main/scala/mesosphere/mesos/TaskBuilder.scala]: Random.nextInt(64510) + 1024 would give the safety of staying within the acceptable port range.

File: src/main/scala/mesosphere/marathon/state/AppDefinition.scala [hunk 4]
```diff
@@ -396,40 +373,41 @@ case class AppDefinition(
...
+          case (appPort, hostPort) =>
+            PortAssignment(Some(appPort.name), effectiveIpAddress, hostPort)
         }.toList
     }
 
     def fromPortMappings: Option[Seq[PortAssignment]] =
       for {
-        pms <- portMappings
+        c <- container
+        pms <- c.portMappings
```
Comment (jasongilanfarr) [inline src/main/scala/mesosphere/marathon/state/AppDefinition.scala]: no idea if this is correct, but can you do "if pms.hostPort.isDefined" here and remove it from the filter? If not, line 387 should be "withFilter()" instead of "filter()"

File: src/main/scala/mesosphere/mesos/TaskBuilder.scala (hunk 1 above)
Comment (jdef) [inline src/main/scala/mesosphere/mesos/TaskBuilder.scala]: from @jasongilanfarr 
> I think we should keep PORTN for legacy/backward compat reasons but use PORT_NAME and maybe PORT_NAME_CP for container ports?

**Comment by jdef** [comment]
Integration testing w/ navstar builds fails. Marathon complains that:
```
Jun 20 05:21:33 ip-10-10-0-202 java[9522]: [2016-06-20 05:21:33,591] WARN mapping exception to status code (mesosphere.marathon.api.MarathonExceptionMapper$$EnhancerByGuice$$a7492536:qtp850429804-35)
Jun 20 05:21:33 ip-10-10-0-202 java[9522]: java.lang.IllegalArgumentException: requirement failed: IP address and ports are not allowed at the same time
Jun 20 05:21:33 ip-10-10-0-202 java[9522]: at scala.Predef$.require(Predef.scala:219)
Jun 20 05:21:33 ip-10-10-0-202 java[9522]: at mesosphere.marathon.state.AppDefinition.<init>(AppDefinition.scala:103)
Jun 20 05:21:33 ip-10-10-0-202 java[9522]: at mesosphere.marathon.state.AppDefinition.copy(AppDefinition.scala:38)
Jun 20 05:21:33 ip-10-10-0-202 java[9522]: at mesosphere.marathon.state.GroupManager$$anonfun$16.apply(GroupManager.scala:271)
Jun 20 05:21:33 ip-10-10-0-202 java[9522]: at mesosphere.marathon.state.GroupManager$$anonfun$16.apply(GroupManager.scala:267)
Jun 20 05:21:33 ip-10-10-0-202 java[9522]: at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:245)
Jun 20 05:21:33 ip-10-10-0-202 java[9522]: at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:245)
Jun 20 05:21:33 ip-10-10-0-202 java[9522]: at scala.collection.immutable.Set$Set2.foreach(Set.scala:111)
```

Application definition:
```
{
 "id": "jdef-networking",
 "cmd": "ip -o addr; sleep 30",
 "cpus": 0.10,
 "mem": 64,
 "instances": 1,
 "backoffFactor": 1.14472988585,
 "backoffSeconds": 5,
 "ipAddress": { "networkName": "overlay-1" },
 "container": {
   "type": "DOCKER",
   "docker": {
     "network": "USER",
     "image": "busybox",
     "portMappings": [{
       "containerPort": 123, "servicePort": 80, "name": "foo3"
     }]
   }
 }
}
```

Suspect that Marathon is injecting default `ports` into the appdef and then complaining that `ports` and `portMappings` can't be defined at the same time.

File: src/main/scala/mesosphere/marathon/api/v2/json/Formats.scala [hunk 5]
```diff
@@ -857,6 +850,26 @@ trait AppAndGroupFormats {
         "secrets" -> runSpec.secrets,
         "taskKillGracePeriodSeconds" -> runSpec.taskKillGracePeriod
       )
+      // top-level ports fields are incompatible with IP/CT
+      if (runSpec.ipAddress.isEmpty) {
+        appJson = appJson ++ Json.obj(
+          // the ports field was written incorrectly in old code if a container was specified
+          // it should contain the service ports
+          //
+          // "ports" -> runSpec.servicePorts, -- can't specify both this and portDefinitions and be valid,
```
Comment (gkleiman) [inline src/main/scala/mesosphere/marathon/api/v2/json/Formats.scala]: Removing this will break mesosphere/marathon-lb (see https://github.com/mesosphere/marathon-lb/blob/d43d1fb9f8c9ed41af9becb5ef8d32897da0aa79/utils.py#L114) and probably other tools as well.

File: src/main/scala/mesosphere/marathon/state/AppDefinition.scala [hunk 6]
```diff
@@ -104,18 +104,6 @@ case class AppDefinition(
 
   lazy val portNumbers: Seq[Int] = portDefinitions.map(_.port)
 
-  /**
-    * Returns true if all health check port index values are in the range
-    * of ths app's ports array, or if defined, the array of container
-    * port mappings.
-    */
-  def portIndicesAreValid(): Boolean = {
```
Comment (gkleiman) [inline src/main/scala/mesosphere/marathon/state/AppDefinition.scala]: Why did you remove this validation?

File: src/main/scala/mesosphere/mesos/TaskBuilder.scala [hunk 7]
```diff
@@ -342,33 +346,83 @@ object TaskBuilder {
...
 
-      definedPorts.zip(assignedPorts).foreach {
-        case (defined, assigned) =>
-          if (defined != AppDefinition.RandomPortValue) {
-            env += (s"PORT_$defined" -> assigned.toString)
+        // matches container-port-only mappings; no host port was defined for this mapping
+        case (None, n) =>
+          definedPorts.lift(n) match {
+            case Some(containerPort) if containerPort == AppDefinition.RandomPortValue =>
+              val cp = ContainerPortGenerator.next
```
Comment (gkleiman) [inline src/main/scala/mesosphere/mesos/TaskBuilder.scala]: Our documentation says:

> Using the value 0 for any port settings indicates to Marathon that you would like a random port assignment. However, if containerPort is set to 0 within a portMapping, it is set to the same value as hostPort.

Wouldn't this new code result in `hostPort != containerPort` if both are `0`?

File: src/main/scala/mesosphere/mesos/TaskBuilder.scala [hunk 8]
```diff
@@ -342,33 +346,83 @@ object TaskBuilder {
...
+              generatedPorts += n -> cp
+              env += (s"PORT$n" -> cp.toString)
+            case Some(containerPort) if containerPort != AppDefinition.RandomPortValue =>
+              env += (s"PORT$n" -> containerPort.toString)
+            case _ => () // should never happen, means that index was out of bounds!
```
Comment (gkleiman) [inline src/main/scala/mesosphere/mesos/TaskBuilder.scala]: So maybe we should log an error or warning if this happens?

File: src/main/scala/mesosphere/mesos/TaskBuilder.scala [hunk 9]
```diff
@@ -342,33 +346,83 @@ object TaskBuilder {
...
+            p
+          }
+          else next // TODO(jdef) **highly** unlikely, but still possible that the port range could be exhausted
+        }
+      }
 
       assignedPorts.zipWithIndex.foreach {
-        case (p, n) =>
+        // matches fixed or dynamic host port assignments
+        case (Some(p), n) =>
```
Comment (gkleiman) [inline src/main/scala/mesosphere/mesos/TaskBuilder.scala]: We should use longer, more verbose variable names to improve readability.

File: src/test/scala/mesosphere/mesos/TaskBuilderTest.scala [hunk 10]
```diff
@@ -1409,6 +1420,93 @@ class TaskBuilderTest extends MarathonSpec with Matchers {
...
+
+    val env: Map[String, String] =
+      taskInfo.getCommand.getEnvironment.getVariablesList.asScala.toList.map(v => v.getName -> v.getValue).toMap
+
+    // port0 is not allocated from the offer since it's container-only, but it should also not
+    // overlap with other (fixed or dynamic) container ports
+    assert(env.contains("PORT0"))
+    val p0 = env("PORT0")
+    assert("0" != env("PORT0"))
+    assert("25003" != env("PORT0"))
```
Comment (gkleiman) [inline src/test/scala/mesosphere/mesos/TaskBuilderTest.scala]: Where does `25003` come from?

**Comment by aquamatthias** [comment]
Thanks!

File: src/main/scala/mesosphere/marathon/state/AppDefinition.scala (hunk 6 above)
Comment (jdef) [inline src/main/scala/mesosphere/marathon/state/AppDefinition.scala]: because it wasn't used

File: src/main/scala/mesosphere/marathon/api/v2/json/Formats.scala [hunk 11]
```diff
@@ -853,16 +853,7 @@ trait AppAndGroupFormats {
...
-          // it should contain the service ports
-          //
-          // "ports" -> runSpec.servicePorts, -- can't specify both this and portDefinitions and be valid,
-          // so I'm removing it. it's been deprecated for a while, users will have to deal.
-          //
-          // TODO(jdef) honestly, it's still very confusing even without IP/CT mode if you're
-          // using BRIDGE networking because portMappings stuff bubbles up to the top-level
-          // ports fields here -- which results in serialized JSON that's invalid according to
-          // our own validation funcs.
+          "ports" -> runSpec.servicePorts,
```
Comment (jdef) [inline src/main/scala/mesosphere/marathon/api/v2/json/Formats.scala]: why was this added back?

File: src/test/scala/mesosphere/marathon/api/v2/AppsResourceTest.scala [hunk 12]
```diff
@@ -356,9 +356,15 @@ class AppsResourceTest extends MarathonSpec with MarathonActorSupport with Match
...
         ))
       )),
-      portDefinitions = Nil
+      portDefinitions = Seq.empty
     )
-    val (body, plan) = prepareApp(app)
+
+    val group = Group(PathId("/"), Set(app))
+    val plan = DeploymentPlan(group, group)
+    val body = Json.stringify(Json.toJson(app).as[JsObject] - "ports").getBytes("UTF-8")
```
Comment (jdef) [inline src/test/scala/mesosphere/marathon/api/v2/AppsResourceTest.scala]: pretty silly that we generate invalid JSON, then correct with some JSON algebra, in order to process it effectively. We **really** shouldn't be generating JSON that doesn't pass our own validation routines. very sad to see this

File: src/main/scala/mesosphere/marathon/api/v2/json/Formats.scala (hunk 5 above)
Comment (jdef) [inline src/main/scala/mesosphere/marathon/api/v2/json/Formats.scala]: so why can't marathon-lb change to use `portDefinitions`?

**Comment by dvpe** [comment]
I am trying to use overlays network and when I use in my descriptor the default is:
```
"portDefinitions": [
    {
      "protocol": "tcp",
      "port": 10001
    }
```
For using overlay network I need:
```
  "ipAddress": {
    "networkName": "dcos-1"
  }
```
The error: IP address and ports are not allowed at the same time
The problem is when ports is empty, by default, it is genering RandomPorts (AppDefinition.scala)

**Comment by jdef** [comment]
Please use the portMappings field instead of portDefinitions. If you're
using virtual networking (USER mode) then you can omit hortPort to avoid
port resource allocation

**Comment by dvpe** [comment]
I try:
`  "ipAddress": {
    "networkName": "dcos-1"
  },
  "residency": null,
  "secrets": null,
  "taskKillGracePeriodSeconds": null,
  "portMappings": [
    {
      "protocol": "tcp",
      "port": 10000,
      "name": "api"
    }
  ],`
The error: _requirement failed: IP address (Some(IpAddress(List(),Map(),DiscoveryInfo(List()),Some(dcos-1)))) and ports (Vector(PortDefinition(10000,tcp,Some(api),Map()))) are not allowed at the same time_


I try:
`  "ipAddress": {
    "networkName": "dcos-1"
  },
  "residency": null,
  "secrets": null,
  "taskKillGracePeriodSeconds": null,`

The error: _requirement failed: IP address (Some(IpAddress(List(),Map(),DiscoveryInfo(List()),Some(dcos-1)))) and ports (Vector(PortDefinition(10000,tcp,Some(api),Map()))) are not allowed at the same time_

**Comment by jdef** [comment]
Did you remove the `portDefinitions` declaration?

**Comment by akamalov** [comment]
@jdef, getting exactly the same problem:

```
{
                        "id": "zk1",
                        "container": {
                                "type": "DOCKER",
                                "docker": {
                                        "image": "confluentinc/cp-zookeeper:latest",
                                        "network": "BRIDGE",
                                        "parameters": [{
                                                "key": "hostname",
                                                "value": "zk1.marathon.mesos"
                                        }, {
                                                "key": "net",
                                                "value": "nuagetestnet"
                                        }],
                                        "portMappings": [{
                                                "containerPort": 2181,
                                                "hostPort": 22181,
                                                "protocol": "tcp",
                                                "servicePort": 10181
                                        }, {
                                                "containerPort": 2888,
                                                "hostPort": 22888,
                                                "protocol": "tcp",
                                                "servicePort": 10888
                                        }, {
                                                "containerPort": 3888,
                                                "hostPort": 23888,
                                                "protocol": "tcp",
                                                "servicePort": 11888
                                        }]
                                },
                                "volumes": []
                        },
                        "env": {
                                "ZOOKEEPER_SERVER_ID": "1",
                                "ZOOKEEPER_TICK_TIME": "2000",
                                "ZOOKEEPER_INIT_LIMIT": "5",
                                "ZOOKEEPER_CLIENT_PORT": "2181",
                                "ZOOKEEPER_SYNC_LIMIT": "2",
                                "ZOOKEEPER_LOG4J_ROOT_LOGLEVEL": "INFO",
                                "ZOOKEEPER_SERVERS": "zk1.marathon.mesos:2888:3888;zk2.marathon.mesos:2888:3888;zk3.marathon.mesos:2888:3888;zk4.marathon.mesos:2888:3888;zk5.marathon.mesos:2888:3888",
                                "NUAGE-DOMAIN": "l3domain",
                                "NUAGE-ZONE": "Zone1",
                                "NUAGE-NETWORK": "Subnet1",
                                "NUAGE-ENTERPRISE": "Test",
                                "NUAGE-USER": "admin"
                        },
                        "cpus": 0.5,
                        "mem": 512,
                        "instances": 1,
                        "maxLaunchDelaySeconds": 10,
                        "constraints": [
                                ["hostname", "UNIQUE"]
                        ],
                        "labels": {
                                "HAPROXY_GROUP": "internal"
                        },
                        "upgradeStrategy": {
                                "minimumHealthCapacity": 0
                        },
                       "ipAddress": {}
 }
```
In my case I am requesting an IP address from Nuage SDN, but the setup displayed above (as-is) fails with the same error:

```
[root@mmaster1 zk.overlay]# curl --header "Authorization: token=$(dcos config show core.dcos_acs_token)" -k -H "Content-Type: application/json" -X POST -d@zk.nuage.single.2.json http://192.160.120.157:8080/v2/apps
{"message":"requirement failed: IP address and ports are not allowed at the same time"}
```
